django/db/models/fields/related_descriptors.py

```diff
@@ -798,6 +798,81 @@ def _get_set_deprecation_msg_params(self):
             self.rel.get_accessor_name() if self.reverse else self.field.name,
         )
 
+    def add_relations(self, relations, *, assert_no_collisions=False):
+        """
+        Creates many (M1, M2) relations with O(1) database queries.
+
+        Any requested relations that already exist will be left alone.
+
+        If you assert that none of the requested relations already exist,
+        you can pass assert_no_collisions=True to save 1 database query.
+        """
+        if len(relations) == 0:
+            return
+
+        (M1_M2, m1_pk_field_name, m2_pk_field_name, m1_m2_pk_tuples) = \
+            self._parse_relations(relations)
+
+        if assert_no_collisions:
+            existing_m1_m2_pk_tuples = set()  # type: Set[Tuple[PK, PK]]
+        else:
+            q = M1_M2.objects.none()
+            for (m1_pk, m2_pk) in m1_m2_pk_tuples:
+                q |= Q(**{m1_pk_field_name: m1_pk, m2_pk_field_name: m2_pk})
+            existing_m1_m2_pk_tuples = {
+                (getattr(m1_m2, m1_pk_field_name), getattr(m1_m2, m2_pk_field_name))
+                for m1_m2 in
+                M1_M2.objects.filter(q)
+            }
+
+        M1_M2.objects.bulk_create([
+            M1_M2(**{m1_pk_field_name: m1_pk, m2_pk_field_name: m2_pk})
+            for (m1_pk, m2_pk) in m1_m2_pk_tuples
+            if (m1_pk, m2_pk) not in existing_m1_m2_pk_tuples
```

Comment:
Could we leverage the `ignore_conflicts` parameter to not require the collisions query on databases that support it?

Reply:
I didn't know about that parameter. I'll do some reading on it...